CSUB-264: Minor follow up

atodorov opened the pull request:
Description of proposed changes:


Practical tips for PR review & merge:

- [ ] All GitHub Actions report PASS
- [ ] Newly added code/functions have unit tests
  - [ ] Coverage tools report all newly added lines as covered
  - [ ] The positive scenario is exercised
  - [ ] Negative scenarios are exercised, e.g. assert on all possible errors
  - [ ] Assert on events triggered if applicable
  - [ ] Assert on changes made to storage if applicable
- [ ] Modified behavior/functions - try to make sure above test items are covered
- [ ] Integration tests are added if applicable/needed

atodorov committed:
Add pre_upgrade/ & post_upgrade hooks in v7

for use with try-runtime and upgrades testing

atodorov commented on pallets/creditcoin/src/migrations/v7.rs:
```diff
@@ -73,6 +99,18 @@ pub mod tests {
+					unreachable!()
+				}
+			};
+			assert_eq!(unverified, migrated_unverified);
```
Currently failing with 

```
error[E0308]: mismatched types
   --> pallets/creditcoin/src/migrations/v7.rs:113:4
    |
113 |             assert_eq!(unverified, migrated_unverified);
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `types::transfer::UnverifiedTransfer`, found struct `types::collect_coins::UnverifiedCollectedCoins`
    |
    = note: expected struct `types::transfer::UnverifiedTransfer<AccountId32, u64, H256, u64>`
               found struct `types::collect_coins::UnverifiedCollectedCoins`
```

CAGS295 commented on pallets/creditcoin/src/migrations/v7.rs:
```diff
@@ -20,6 +20,21 @@ pub(crate) fn migrate<T: Config>() -> Weight {
+#[cfg(feature = "try-runtime")]
+pub(crate) fn post_upgrade<T: Config>() -> Result<(), &'static str> {
+	ensure!(
+		StorageVersion::get::<crate::Pallet<T>>() == 7,
```
It looks trivial to me, we always bump storage version at the end of a successful migration. Why would it not be 7?
 We will need to modify this line on each storage version bump (i.e. for each migration), should we include a check in CI?

atodorov replied:
> It looks trivial to me, we always bump storage version at the end of a successful migration. Why would it not be 7?

For example a copy& paste error in `migrations/mod.rs` which incorrectly sets storage version. 

IIRC we didn't come up with anything more to assert on here other then the number. The rest of the testing activities will be around making sure migrations are applied without crashing and then forking the blockchain and running the integration test suite against the fork.


> We will need to modify this line on each storage version bump (i.e. for each migration), should we include a check in CI?

Good point. I am not sure how to handle this elegantly though. Will think about it a bit more.

atodorov replied:
@CAGS295 added a new sanity check script, see https://github.com/gluwa/creditcoin/actions/runs/3874453271/jobs/6605673537

atodorov committed:
Remove unused scripts/init.sh

atodorov committed:
Add script which will assert on migrations

atodorov committed:
Add the missing post_upgrade() and StorageVersion assertions

which will satisfy check-storage-version-assert-in-migrations.sh

CAGS295 committed:
move difficulty primitive

AdaJane commented on scripts/check-storage-version-assert-in-migrations.sh:
```diff
@@ -0,0 +1,23 @@
+    grep "fn post_upgrade" "$MIGRATION" || fail "fn post_upgrade() not found in $MIGRATION"
+
+    # check if the migration file asserts on storage version
+    grep "StorageVersion::get::<crate::Pallet<T>>() == $VERSION" "$MIGRATION" || fail "Assertion 'StorageVersion == $VERSION' not found in $MIGRATION"
```
Does this script provide any needed functionality? I'm generally not a fan of testing for an exact line of code as opposed to testing for the intended functionality.

I don't know the history here, but it seems like this is a very fragile script that doesn't help too much.

atodorov replied:
The purpose of this script is to make sure that every migration contains a `post_upgrade()` hook and that each hook compares the the StorageVersion number (based on the migration version).  

The `post_upgrade()`  function is a building block in our strategy to test runtime upgrades and currently the cause of failure seen here: https://github.com/gluwa/creditcoin/actions/runs/3902689057/jobs/6666466133. For more context around the current state of testing upgrades see the last commit in https://github.com/gluwa/creditcoin/pull/816. 

If we forget the function or to assert on the correct storage version number the migration process will not fail in CI, even if something went wrong,  and we will miss the alert.

atodorov commented on pallets/creditcoin/src/migrations/v7.rs:
```diff
@@ -20,6 +22,23 @@ pub(crate) fn migrate<T: Config>() -> Weight {
+}
+
+#[cfg(feature = "try-runtime")]
+#[allow(dead_code)]
```
Interestingly enough Clippy doesn't complain about the same functions in `v6.rs`. 

@nathanwhit any idea why that is?

atodorov replied:
FTR https://github.com/gluwa/creditcoin/blob/dev/pallets/creditcoin/src/migrations/mod.rs#L13-L28

I will look more into it.

AdaJane reviewed (dismissed):
I still wonder if there's a better way to ensure the upgrade hooks are implemented (and correctly), but since the shell script is in direct response to a prior issue I don't think it's worth holding this up.

LGTM

atodorov committed:
Redesign how pre_upgrade()/post_upgrade() functions are called

these are no longer Substrate hooks but rather a place where we can
perform more assertions while testing migrations in CI.

scripts/check-storage-version-assert-in-migrations.sh now becomes
useless.

AdaJane commented on pallets/creditcoin/src/migrations/mod.rs:
```diff
@@ -9,55 +9,79 @@ mod v5;
-	Ok(())
-}
-
 pub(crate) fn migrate<T: Config>() -> Weight {
```
This definitely seems like an improvement over the shell script. What do you think about cleaning this up and making it a bit more maintanable by doing something like this:

```
use std::collections::HashMap
pub(crate) fn migrate<T: Config>() -> Weight {
	let version = StorageVersion::get::<Pallet<T>>();
	let mut weight: Weight = Weight::zero();

	let pre_hooks = HashMap::from([
		(1, v1::pre_upgrade::<T>),
		(2, v2::pre_upgrade::<T>),
		(3, v3::pre_upgrade::<T>),
		(4, v4::pre_upgrade::<T>),
		(5, v5::pre_upgrade::<T>),
		(6, v6::pre_upgrade::<T>),
		(7, v7::pre_upgrade::<T>),
	]);

	let migrations = HashMap::from([
		(1, v1::migrate::<T>),
		(2, v2::migrate::<T>),
		(3, v3::migrate::<T>),
		(4, v4::migrate::<T>),
		(5, v5::migrate::<T>),
		(6, v6::migrate::<T>),
		(7, v7::migrate::<T>),
	]);

	let post_hooks = HashMap::from([
		(1, v1::post_upgrade::<T>),
		(2, v2::post_upgrade::<T>),
		(3, v3::post_upgrade::<T>),
		(4, v4::post_upgrade::<T>),
		(5, v5::post_upgrade::<T>),
		(6, v6::post_upgrade::<T>),
		(7, v7::post_upgrade::<T>),
	]);

	// Attempt to retrieve the pre_hook based on version
	let pre_hook = pre_hooks.get(version.into())
		.expect(&format!("No pre_upgrade hook for version {:?}", version));

	#[cfg(feature = "try-runtime")]
	pre_hook();

	let migrate = migrations.get(version.into())
		.expect(&format!("No Migration hook for version {:?}", version));

	weight.saturating_accrue(migrate());
	StorageVersion::new(version.into()).put::<Pallet<T>>();

	let post_hook = post_hooks.get(version.into())
		.expect(&format!("No post_upgrade hook for version {:?}", version));

	#[cfg(feature = "try-runtime")]
	post_hook();

	weight
}
```

atodorov replied:
Hmm, that looks neater but migrations will not be applied exactly as before. For example: 

if I am on v5 and updating to latest, which is v7 right now, I want both v6 and v7 applied. 

The example above will actually try to apply v5 (b/c that's what's in storage).  

Maybe an loop over [version+1, length_of_migrations] ? How bad could that be @CAGS295, @nathanwhit ? 

OTOH I do prefer the explicitness of these `if version <` blocks. Maybe combine the explicit if conditions with a helper function which executes the pre-migrate-post trio? WDYT ?

atodorov committed:
Update migrate() to use a loop & a map of callbacks

makes the code more concise and easier to read and removes the explicit
checks for version number before applying the next migration!

atodorov commented on pallets/creditcoin/src/migrations/mod.rs:
```diff
@@ -9,55 +10,33 @@ mod v5;
+		(5, (v5::pre_upgrade::<T>, v5::migrate::<T>, v5::post_upgrade::<T>)),
+		(6, (v6::pre_upgrade::<T>, v6::migrate::<T>, v6::post_upgrade::<T>)),
+		(7, (v7::pre_upgrade::<T>, v7::migrate::<T>, v7::post_upgrade::<T>)),
+	]);
```
I tried following a suggestion from Carlos but can't seem to figure out the compiler errors.

CAGS295 replied:
try [this](https://github.com/gluwa/creditcoin/tree/feature/pre-post-migration-checks) impl. It is more evident here that the version check in post_upgrade is trivial.

CAGS295 committed:
Migration trait

atodorov committed:
Bump version

Commit:
Auto-update creditcoin-js type definitions

AdaJane reviewed (dismissed):
This updated implementation seems like a big improvement.

atodorov commented on pallets/creditcoin/src/migrations/mod.rs:
```diff
@@ -9,55 +17,30 @@ mod v5;
+			StorageVersion::new(migration_idx).put::<Pallet<T>>();
+			#[cfg(feature = "try-runtime")]
+			calls.post_upgrade();
+		}
```
Looks good, we can see both v6 and v7 in the logs here https://github.com/gluwa/creditcoin/actions/runs/3940108310/jobs/6741258719 (that's from #816).

CAGS295 reviewed (dismissed):
Not sure if my review is valid but here you go.

nathanwhit commented on pallets/creditcoin/src/migrations/v1.rs:
```diff
@@ -181,90 +182,113 @@ type DealOrders<T: crate::Config> = StorageDoubleMap<
+	pub(super) fn new() -> Self {
+		Self(PhantomData::<Runtime>)
+	}
+}
```
Nitpick but a weird mix of visibilities:
- struct is `pub(crate)`
- field is `pub`
- `new` is `pub(super)`

I think the field can be private since it's not meant to be accessed, and the struct could be `pub(super)` (unless we need to name it elsewhere in the crate, but not sure why we would)

nathanwhit commented on pallets/creditcoin/src/migrations/v2.rs:
```diff
@@ -97,57 +97,76 @@ type Transfers<T: crate::Config> = StorageMap<
-pub(crate) fn migrate<T: Config>() -> Weight {
-	let mut weight: Weight = Weight::zero();
-	let weight_each = T::DbWeight::get().reads_writes(1, 1);
+pub(crate) struct Migration<Runtime>(pub PhantomData<Runtime>);
```
Same visibilities comment applies

nathanwhit commented on pallets/creditcoin/src/migrations/v2.rs:
```diff
@@ -157,6 +176,7 @@ mod test {
 		tests::TestInfo,
 		DealOrderId, DoubleMapExt, Duration, OfferId, TransferId,
 	};
+	use core::convert::TryInto;
```
Not sure why this snuck in

nathanwhit commented on pallets/creditcoin/src/migrations/v3.rs:
```diff
@@ -148,64 +149,85 @@ type DealOrders<T: crate::Config> = StorageDoubleMap<
-pub(crate) fn migrate<T: Config>() -> Weight {
-	let mut weight: Weight = Weight::zero();
-	let weight_each = T::DbWeight::get().reads_writes(1, 1);
+pub(crate) struct Migration<Runtime>(pub PhantomData<Runtime>);
```
Weird visibilities again

nathanwhit commented on pallets/creditcoin/src/migrations/v2.rs:
```diff
@@ -97,57 +97,76 @@ type Transfers<T: crate::Config> = StorageMap<
-		|_id, transfer| {
+impl<Runtime> Migration<Runtime> {
+	pub(super) fn new() -> Self {
+		Self(PhantomData::<Runtime>)
```
```suggestion
		Self(PhantomData)
```

nathanwhit commented on pallets/creditcoin/src/migrations/v1.rs:
```diff
@@ -181,90 +182,113 @@ type DealOrders<T: crate::Config> = StorageDoubleMap<
-	});
+impl<Runtime: Config> Migration<Runtime> {
+	pub(super) fn new() -> Self {
+		Self(PhantomData::<Runtime>)
```
```suggestion
		Self(PhantomData)
```
As much as I love a good [turbofish](https://doc.rust-lang.org/book/appendix-02-operators.html?highlight=turbofish#non-operator-symbols), it's not necessary here (and it's kinda noisy syntax)

nathanwhit commented on pallets/creditcoin/src/migrations/v3.rs:
```diff
@@ -148,64 +149,85 @@ type DealOrders<T: crate::Config> = StorageDoubleMap<
-	);
+impl<Runtime> Migration<Runtime> {
+	pub(super) fn new() -> Self {
+		Self(PhantomData::<Runtime>)
```
```suggestion
		Self(PhantomData)
```

nathanwhit commented on pallets/creditcoin/src/migrations/v4.rs:
```diff
@@ -24,15 +28,36 @@ pub struct Address<AccountId> {
-pub(crate) fn migrate<T: Config>() -> Weight {
-	let sp_io::MultiRemovalResults { unique: count_removed, .. } =
-		Addresses::<T>::clear(u32::MAX, None);
+pub(crate) struct Migration<Runtime>(pub PhantomData<Runtime>);
```
visibilities

nathanwhit commented on pallets/creditcoin/src/migrations/v4.rs:
```diff
@@ -24,15 +28,36 @@ pub struct Address<AccountId> {
-	T::DbWeight::get().writes(count_removed.saturated_into())
+impl<Runtime> Migration<Runtime> {
+	pub(super) fn new() -> Self {
+		Self(PhantomData::<Runtime>)
```
```suggestion
		Self(PhantomData)
```

nathanwhit commented on pallets/creditcoin/src/migrations/v7.rs:
```diff
@@ -1,27 +1,49 @@
-			Task::CollectCoins(pending) => TaskV2::<T>::to_id(pending),
-			Task::VerifyTransfer(pending) => TaskV2::<T>::to_id(pending),
-		};
+pub(crate) struct Migration<Runtime>(pub PhantomData<Runtime>);
```
visibilities

nathanwhit commented on pallets/creditcoin/src/migrations/v7.rs:
```diff
@@ -1,27 +1,49 @@
-		T::TaskScheduler::insert(&k1, &id, v);
+impl<Runtime> Migration<Runtime> {
+	pub(super) fn new() -> Self {
+		Self(PhantomData::<Runtime>)
```
```suggestion
		Self(PhantomData)
```

nathanwhit reviewed (dismissed):
A bunch of nitpicks but nothing big or blocking

atodorov committed:
Adjust visibility
